python/ray/data/_internal/block_batching.py

```diff
@@ -88,7 +88,8 @@ def get_batches(block: Optional[ObjectRef[Block]] = None) -> Iterator[BatchType]
             # NOTE: Since we add one block at a time and then immediately consume
             # batches, we don't need to check batcher.can_add() before adding the block;
             # it will always be True, and batcher.add() will assert this internally.
-            batcher.add(block)
+            if BlockAccessor.for_block(block).num_rows() > 0:
+                batcher.add(block)
```

Comment:
Should the `Batcher` handle this internally? 🤔

Reply:
I think that the better fix might be changing [this check in the `Batcher`](https://github.com/ray-project/ray/blob/08c09c5a2305965a463523ef9dbed883bb8a86e2/python/ray/data/_internal/batcher.py#L87-L89) to be
```python
        return self._buffer_size > 0 and (
            self._batch_size is None or self._buffer_size >= self._batch_size
        )
```

Reply:
@clarkzinzow - yeah I found this too, I was planning to fix this part in https://github.com/ray-project/ray/pull/29289, but we can fix in this PR as well.

Change `has_batch()` itself is not enough, because then we still allow add empty block into buffer, and [`next_batch()` will throw exception here](https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/batcher.py#L104).

Example:

```py
batcher.add(empty_block)
batcher.add(non_empty_block)
batcher.next_batch() # throw exception because len(self._buffer) == 2
```

Reply:
Yeah that would need to be modified as well, but I think that having the `Batcher` handle this instead of the caller has a few advantages:
- Empty block handling is pushed down into a lower abstraction level, so the caller doesn't have to worry about it, yielding simpler caller code.
- The `Batcher` can use those empty blocks to construct more meaningful empty batches (with e.g. well-defined schema) when no non-empty blocks are added.

Reply:
To get the benefits of (2), I think we'd need to:
- change that `.has_batch()` check as above,
- in `.next_batch()` for the `batch_size is None` case, filter out all empty blocks in the buffer, assert that there's only one non-empty block in the buffer, return that block,
- if all blocks in the buffer are empty, return a block from `.next_batch()` that has a well-defined schema.

Hmm however, I don't think it would line up with our `Batcher.has_any()` semantics, which only returns `True` if there's data in the buffer... Maybe we'll need to think about this some more.

Reply:
@clarkzinzow - yeah agree we should revisit adding empty block later, it's more tricky.